Libraries/LibWeb/HTML/Navigable.cpp

```diff
@@ -1165,16 +1165,24 @@ WebIDL::ExceptionOr<void> Navigable::populate_session_history_entry_document(
         //  - FIXME: navigationParams's reserved environment is non-null and the result of checking a navigation response's adherence to its embedder policy given navigationParams's response, navigable, and navigationParams's policy container's embedder policy is false; or
-        //  - FIXME: the result of checking a navigation response's adherence to `X-Frame-Options` given navigationParams's response, navigable, navigationParams's policy container's CSP list, and navigationParams's origin is false,
-        if (navigation_params.has<NullOrError>()) {
+        //  - the result of checking a navigation response's adherence to `X-Frame-Options` given navigationParams's response, navigable, navigationParams's policy container's CSP list, and navigationParams's origin is false,
+        if (navigation_params.has<NullOrError>() || !navigation_params.get<GC::Ref<NavigationParams>>()->check_a_navigation_responses_adherence_to_x_frame_options(this)) {
             // 1. Set entry's document state's document to the result of creating a document for inline content that doesn't have a DOM, given navigable, null, and navTimingType. The inline content should indicate to the user the sort of error that occurred.
-            auto error_message = navigation_params.get<NullOrError>().value_or("Unknown error"sv);
+            auto error_message = navigation_params.has<NullOrError>() ? navigation_params.get<NullOrError>().value_or("Unknown error"sv) : "The request was denied."sv;
 
             auto error_html = load_error_page(entry->url(), error_message).release_value_but_fixme_should_propagate_errors();
-            entry->document_state()->set_document(create_document_for_inline_content(this, navigation_id, [error_html](auto& document) {
+            entry->document_state()->set_document(create_document_for_inline_content(this, navigation_id, [error_html, this](auto& document) {
                 auto parser = HTML::HTMLParser::create(document, error_html, "utf-8"sv);
                 document.set_url(URL::URL("about:error"));
                 parser->run();
+
+                // NOTE: Once the page has been set up, the user agent must act as if it had stopped parsing.
+                // FIXME: Directly calling parser->the_end results in a deadlock, because it waits for the warning image to load.
+                //        However the response is never processed when parser->the_end is called.
+                //        Queuing a global task is a workaround for now.
+                queue_a_task(Task::Source::Unspecified, HTML::main_thread_event_loop(), document, GC::create_function(heap(), [parser, &document]() {
+                    parser->the_end(document, nullptr);
```

Comment:
HTMLParser::the_end() is static. Are you sure we don't want to pass the parser in as the second argument? That might avoid the problem here of needing an extra task. iirc the only reason to pass in nullptr is if you're doing fragment parsing(?)

Reply:
[This note](https://github.com/LadybirdBrowser/ladybird/blob/f1bbba2ba54913876830c1de2a293d6faee20535/Libraries/LibWeb/HTML/Parser/HTMLParser.cpp#L253) states that "act as if the user agent had stopped parsing document" means calling the method without a parser. It looks as though the parser is also reset when it is passed. But that wouldn't matter here, as the parser is no longer used anyway. Additionally, you need an extra task in any case.

Reply:
Hm. all right. In that case, you don't need to capture the parser at all. In either case, you should be calling HTMLParser::the_end() by name instead of as a method. calling static methods as if they are object methods "should be" a smell that our linters complain about 🤔